bothan-core/src/registry/validate.rs

```diff
@@ -0,0 +1,52 @@
+use std::collections::HashMap;
+
+use thiserror::Error;
+
+use crate::registry::Registry;
+
+#[derive(Clone, Debug, Error, PartialEq)]
+pub enum ValidationError {
+    #[error("Signal {0} contains a cycle")]
+    CycleDetected(String),
+
+    #[error("Signal {0} contains an invalid dependency")]
+    InvalidDependency(String),
+}
+
+pub enum VisitState {
+    InProgress,
+    Complete,
+}
+
+pub(crate) fn dfs(
```

Comment:
name isDAG mai

Reply:
isDAG implies bool result, will change to validate_signal instead